File: scripts/migrate_users.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';

// Importamos as opções do nosso próprio app para que o script saiba a qual projeto se conectar.
import '../lib/firebase_options.dart';

Future<void> main() async {
  print('Iniciando script de migração...');
  
  // Inicializa a conexão com o Firebase usando as credenciais do seu app
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );

  final firestore = FirebaseFirestore.instance;
  final usersRef = firestore.collection('users');
  int updatedCount = 0;

  try {
    print('Buscando todos os usuários...');
    final querySnapshot = await usersRef.get();
    
    if (querySnapshot.docs.isEmpty) {
      print('Nenhum usuário encontrado. Nada a fazer.');
      return;
    }

    print('Encontrados ${querySnapshot.docs.length} usuários. Verificando...');

    // Usamos um batch para fazer todas as atualizações de uma só vez (mais eficiente)
    final batch = firestore.batch();

    for (final doc in querySnapshot.docs) {
      final data = doc.data();
      // Verifica se o campo 'hasCompletedOnboarding' NÃO existe
      if (!data.containsKey('hasCompletedOnboarding')) {
        print('Atualizando usuário: ${doc.id} (${data['username']})');
        // Adiciona a operação de atualização ao batch
        batch.update(doc.reference, {'hasCompletedOnboarding': true});
        updatedCount++;
      }
    }

    if (updatedCount > 0) {
      print('\nEncontrados $updatedCount usuários antigos. Salvando atualizações...');
      await batch.commit();
      print('Migração concluída com sucesso! $updatedCount perfis foram atualizados.');
    } else {
      print('\nNenhum usuário precisava de atualização. Tudo certo!');
    }

  } catch (e) {
    print('Ocorreu um erro durante a migração: $e');
  }
}

File: scripts/migrate_users.py
import firebase_admin
from firebase_admin import credentials, firestore

# --- CONFIGURAÇÃO ---
cred = credentials.Certificate("scripts/serviceAccountKey.json")
try:
    firebase_admin.get_app()
except ValueError:
    firebase_admin.initialize_app(cred)

db = firestore.client()
print("Conectado ao projeto Firebase com sucesso.")
# --- FIM DA CONFIGURAÇÃO ---

def migrate_users():
    """
    Este script percorre todos os documentos na coleção 'users'.
    Para cada usuário, ele verifica se o campo 'hasCompletedOnboarding' existe.
    Se não existir, ele o adiciona com o valor 'true'.
    """
    users_ref = db.collection('users')
    all_users = users_ref.stream()
    
    batch = db.batch()
    updated_count = 0
    checked_count = 0
    
    print("Iniciando a verificação dos usuários...")

    for user_doc in all_users:
        checked_count += 1
        user_data = user_doc.to_dict()
        
        if 'hasCompletedOnboarding' not in user_data:
            print(f"  -> Usuário {user_doc.id} ({user_data.get('username', 'N/A')}) será atualizado.")
            
            # CORREÇÃO: O método correto é .document(), não .doc()
            user_ref = users_ref.document(user_doc.id) 
            batch.update(user_ref, {'hasCompletedOnboarding': True})
            updated_count += 1
            
            if updated_count > 0 and updated_count % 499 == 0:
                print(f"\n--- Lote de {updated_count} atualizações atingido, enviando para o Firestore... ---\n")
                batch.commit()
                batch = db.batch()

    if updated_count > 0:
        print("\nEnviando o lote final...")
        batch.commit()
        print("Lote final enviado com sucesso!")
    
    print("\n--- Migração Concluída ---")
    print(f"Total de usuários verificados: {checked_count}")
    print(f"Total de usuários atualizados: {updated_count}")

if __name__ == '__main__':
    migrate_users()
